pkg/template/kurl_context.go

```diff
@@ -0,0 +1,197 @@
+package template
+
+import (
+	"fmt"
+	"reflect"
+	"sort"
+	"strconv"
+	"strings"
+	"text/template"
+
+	kurlclientset "github.com/replicatedhq/kurl/kurlkinds/client/kurlclientset/typed/cluster/v1beta1"
+	kurlv1beta1 "github.com/replicatedhq/kurl/kurlkinds/pkg/apis/cluster/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8sconfig "sigs.k8s.io/controller-runtime/pkg/client/config"
+)
+
+func GetKurlValues(installerName, nameSpace string) (*kurlv1beta1.Installer) {
+
+	cfg, err := k8sconfig.GetConfig()
+
+	if err != nil {
+		return nil
+	}
+
+	clientset := kurlclientset.NewForConfigOrDie(cfg)
+
+	installers := clientset.Installers(nameSpace)
+
+	retrieved, err := installers.Get(installerName, metav1.GetOptions{})
+
+	if err != nil {
+		fmt.Printf("Could not retrieve installer crd object, does an the installer '%s' exist in the '%s namespace'?\n", installerName, nameSpace)
+		return nil
+	}
+
+	return retrieved
+}
+
+func NewKurlContext(installerName, nameSpace string) (*KurlCtx) {
+	ctx := &KurlCtx{
+		KurlValues: make(map[string]interface{}),
+	}
+
+	retrieved := GetKurlValues(installerName, nameSpace)
+
+	if retrieved != nil {
+		ctx.AddValuesToKurlContext(retrieved)
+	}
+
+	return ctx
+}
+
+func (ctx KurlCtx) AddValuesToKurlContext(retrieved *kurlv1beta1.Installer) {
+	Spec := reflect.ValueOf(retrieved.Spec)
+
+	for i := 0; i < Spec.NumField(); i++ {
+		Category := reflect.ValueOf(Spec.Field(i).Interface())
+
+		TypeOfCategory := Category.Type()
+
+		RawCategoryName := Category.String()
+		TrimmedRight := strings.Split(RawCategoryName, ".")[1]
+		CategoryName := strings.Split(TrimmedRight, " ")[0]
+
+		for i := 0; i < Category.NumField(); i++ {
+			if Category.Field(i).CanInterface() {
+				ctx.KurlValues[CategoryName+"."+TypeOfCategory.Field(i).Name] = Category.Field(i).Interface()
+			}
+		}
+	}
+}
+
+type KurlCtx struct {
+	KurlValues map[string]interface{}
+}
+
+func (ctx KurlCtx) FuncMap() template.FuncMap {
+	return template.FuncMap{
+		"KurlString": ctx.kurlString,
+		"KurlInt":    ctx.kurlInt,
+		"KurlBool":   ctx.kurlBool,
+		"KurlOption": ctx.kurlOption,
+		"KurlAll":    ctx.kurlAll,
+	}
+}
+
+func (ctx KurlCtx) kurlBool(yamlPath string) bool {
+	if len(ctx.KurlValues) == 0 {
+		fmt.Printf("No Installer CRD has been found, is this a kURL cluster?\n")
```

Comment:
I think this is too verbose.  Same app can be installed on kurl and non-kurl clusters, so this is going to be just noise almost always.

Reply:
changed